### ISSUE TITLE
Find with populate results in CPU getting pegged and node taking up 1.5GB+ of RAM

### ISSUE THREAD

**Issue by blah238** [balderdashy/sails-postgresql#141]
I have yet to isolate the cause, but I'm seeing crazy memory and CPU consumption on a simple `Model.find().populate('field')` or `.populateAll()` when Model gets to be over 500 records or so and ~1500 on the associated model. My PostgreSQL database in its entirety is only ~50MB, but doing the above causes the node process to max out at 25% (quad-core) and balloon to 1.5GB+ of RAM. The PostgreSQL query is done in no time, but Waterline's processing of the results takes ages (~12 minutes) to run. I'm using sails@0.11.0, waterline@0.10.18 and sails-postgresql@0.10.14 running on node.js 0.10.30 32-bit on Windows 7 SP1 64-bit with PostgreSQL 9.4 (x86).

If I take out the `.populate` it's fine. If I use sails-disk instead of sails-postgresql, the CPU usage remains high but the memory usage stays under 250MB and the query is returned within 2-3 seconds).

Will add more info as I can, but just wondering if anyone else has seen this behavior with non-trivial dataset sizes, and what if anything they were able to do to resolve it. I could try 64-bit Node but this has always been problematic on Windows, and I really don't think it should be necessary.

**Comment by devinivy** [comment]
Could live [here](https://github.com/balderdashy/sails-postgresql/blob/master/lib/adapter.js#L506-L507) or in waterline-cursor.

**Comment by devinivy** [comment]
@blah238 does it seem like you're having the same issue as in #141?

**Comment by blah238** [comment]
@devinivy Maybe, but I'm not sure how to confirm it. I see that PR https://github.com/balderdashy/waterline-sequel/pull/15 might address #116. I might try merging it locally tomorrow.

**Comment by blah238** [comment]
@devinivy Unfortunately no change with @Benny-'s patch. I also tried waterline-sequel 0.2.0 and no change.

**Comment by blah238** [comment]
I am able to reproduce this with a toy application:

**api/models/User.js:**
```js
module.exports = {
  attributes: {
    pets: {
      collection: 'pet',
      via: 'owner'
    }
  }
};
```
**api/models/Pet.js:**
```
module.exports = {
  attributes: {
    owner: {
      model: 'user'
    }
  }
};
```

**api/controllers/PetController.js**:
```js
module.exports = {
  find: function (req, res) {
    Pet.find().populate('owner')
    .then(res.ok)
    .catch(res.serverError);
  }
};
```

**config/bootstrap.js**:
```js
module.exports.bootstrap = function(cb) {
  async.eachSeries(_.range(0, 3), function (userIndex, innerCb) {
    User.create().exec(function (err, user) {
      if (err) {
        return innerCb(err);
      }
      async.eachSeries(_.range(0, 100), function (petIndex, innerInnerCb) {
        Pet.create({ owner: user.id }).exec(innerInnerCb);
      }, innerCb);
    });
  }, cb);
};
```

**config/models.js**:
```js
module.exports.models = {
  connection: 'postgresql',
  migrate: 'drop'
};
```

Even with just 3 Users and 300 Pets (100 per User), hitting the `GET /pet` route results in memory leaking at a very quick rate. After 10 hits I'm over 500MB on the node process.

With 3 Users and 3000 Pets (1000 per User), hitting `GET /pet` once results in 25% CPU usage and ~1.5GB of RAM after a couple minutes of waiting for the response.

**Comment by devinivy** [comment]
@dmarcelino @tjwebb mind helping to organize folks to get this fixed-up?  Pretty mission-critical!  I'm willing to help when I find time.  I'm wondering if this is an issue with waterline-cursor or waterline-sequel.

**Comment by tjwebb** [comment]
I think the first step is to test against other adapters and see if there's a similar occurrence. 

I haven't looked at the pg adapter in awhile; it's possible that it uses waterline-cursor even though the db has intrinsic join ability. If that's the case, the pg adapter is probably due for an upgrade in this respect. 

> Waterline's processing of the results takes ages (~12 minutes) to run

If waterline-cursor is being used here, I've seen it be slow, but nowhere near *that* slow. That's crazy. Is it possible nodejs Buffers are implemented differently (and horribly) on Windows? I think a test on linux is also in order.

@blah238 is there any way you can post the datasets you are joining?

**Comment by blah238** [comment]
@devinivy Thanks and yes I agree this is a critical issue, especially if it occurs on other platforms and adapters.

@tjwebb See my latest post, I have reproduced it with just a simple many-to-one association with no other attributes. I can probably test on a Linux VM but it's not really my strong suit. It would probably be better to get someone else to reproduce it, although I have reproduced on two different Windows machines.

I've been playing around with node-inspector's profiler and node-theseus to try and get an idea of what's going on, but again not really sure what I'm doing.

I did notice a few calls being made a large number of times, if that helps to narrow it down at all:

- https://github.com/balderdashy/sails-postgresql/blob/v0.10.14/lib/adapter.js#L657
- https://github.com/balderdashy/waterline-cursor/blob/0.0.5/cursor/cursor.js#L111
- https://github.com/lodash/lodash/blob/2.4.1/dist/lodash.js#L747

**Comment by blah238** [comment]
@devinivy @tjwebb just tested with sails-mysql and seeing the exact same behavior.

**Comment by dmarcelino** [comment]
@blah238, thanks for submitting the code and examples, that helps. I've created a load test ([`8c2ff33`](https://github.com/balderdashy/waterline-adapter-tests/commit/8c2ff3388dbecc105678efaf73c92f602bbfd92b)) for waterline and it's adapters. Please check it out, namely [runner.js](https://github.com/balderdashy/waterline-adapter-tests/blob/8c2ff3388dbecc105678efaf73c92f602bbfd92b/test/load/runner.js) to see if I didn't mess it somehow. Anyway, running these tests never takes more than 18MB (measured using `process.memoryUsage()`) on Mac/Linux. You can see the travis build results:
https://travis-ci.org/balderdashy/waterline-adapter-tests/jobs/56555832#L608

Can you try running this to see if your results differ? Thanks

**Comment by blah238** [comment]
@dmarcelino Thanks! Here you go (I disabled the mongo and redis tests since I don't have those installed). I noticed that the node process was using upwards of 500MB while running the sails-postgresql and sails-mysql tests, whlie the sails-disk and sails-memory tests only took up ~100MB.

```
$ node test/load


-------------------------------------------------------------------------------------------
                                     sails-postgresql
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 37.75 MB, heapTotal: 35.53 MB, heapUsed: 31.28 MB
.........
20: rss: 37.58 MB, heapTotal: 35.22 MB, heapUsed: 30.58 MB
.........
30: rss: 37.59 MB, heapTotal: 35.22 MB, heapUsed: 29.77 MB
.........
40: rss: 39.77 MB, heapTotal: 37.35 MB, heapUsed: 34.24 MB
.........
50: rss: 38.13 MB, heapTotal: 35.83 MB, heapUsed: 31.64 MB
.........
60: rss: 38.29 MB, heapTotal: 35.93 MB, heapUsed: 31.51 MB
.........
70: rss: 38.51 MB, heapTotal: 36.14 MB, heapUsed: 31.73 MB
.........
80: rss: 38.63 MB, heapTotal: 36.24 MB, heapUsed: 31.86 MB
.........
90: rss: 38.75 MB, heapTotal: 36.44 MB, heapUsed: 32.34 MB
.........
100: rss: 35.95 MB, heapTotal: 33.60 MB, heapUsed: 24.63 MB
collections teardown done
time elapsed: 115076ms
closing code: 0


-------------------------------------------------------------------------------------------
                                     sails-memory
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 15.00 MB, heapTotal: 13.69 MB, heapUsed: 12.22 MB
.........
20: rss: 15.00 MB, heapTotal: 13.69 MB, heapUsed: 12.22 MB
.........
30: rss: 15.00 MB, heapTotal: 13.69 MB, heapUsed: 12.22 MB
.........
40: rss: 15.01 MB, heapTotal: 13.69 MB, heapUsed: 12.23 MB
.........
50: rss: 14.99 MB, heapTotal: 13.69 MB, heapUsed: 12.23 MB
.........
60: rss: 14.99 MB, heapTotal: 13.69 MB, heapUsed: 12.23 MB
.........
70: rss: 14.99 MB, heapTotal: 13.69 MB, heapUsed: 12.23 MB
.........
80: rss: 14.99 MB, heapTotal: 13.69 MB, heapUsed: 12.23 MB
.........
90: rss: 14.99 MB, heapTotal: 13.69 MB, heapUsed: 12.23 MB
.........
100: rss: 14.99 MB, heapTotal: 13.69 MB, heapUsed: 12.23 MB
collections teardown done
time elapsed: 4616ms
closing code: 0


-------------------------------------------------------------------------------------------
                                     sails-disk
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 14.13 MB, heapTotal: 12.77 MB, heapUsed: 11.46 MB
.........
20: rss: 14.13 MB, heapTotal: 12.77 MB, heapUsed: 11.46 MB
.........
30: rss: 14.13 MB, heapTotal: 12.77 MB, heapUsed: 11.46 MB
.........
40: rss: 14.11 MB, heapTotal: 12.77 MB, heapUsed: 11.46 MB
.........
50: rss: 14.11 MB, heapTotal: 12.77 MB, heapUsed: 11.46 MB
.........
60: rss: 14.11 MB, heapTotal: 12.77 MB, heapUsed: 11.47 MB
.........
70: rss: 14.11 MB, heapTotal: 12.77 MB, heapUsed: 11.47 MB
.........
80: rss: 14.11 MB, heapTotal: 12.77 MB, heapUsed: 11.47 MB
.........
90: rss: 14.11 MB, heapTotal: 12.77 MB, heapUsed: 11.47 MB
.........
100: rss: 14.11 MB, heapTotal: 12.77 MB, heapUsed: 11.47 MB
collections teardown done
time elapsed: 4669ms
closing code: 0


-------------------------------------------------------------------------------------------
                                     sails-mysql
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 39.01 MB, heapTotal: 36.64 MB, heapUsed: 33.66 MB
.........
20: rss: 39.22 MB, heapTotal: 36.85 MB, heapUsed: 33.06 MB
.........
30: rss: 30.68 MB, heapTotal: 28.52 MB, heapUsed: 24.66 MB
.........
40: rss: 32.45 MB, heapTotal: 30.14 MB, heapUsed: 17.03 MB
.........
50: rss: 36.41 MB, heapTotal: 34.00 MB, heapUsed: 21.20 MB
.........
60: rss: 36.11 MB, heapTotal: 33.80 MB, heapUsed: 31.81 MB
.........
70: rss: 44.00 MB, heapTotal: 41.42 MB, heapUsed: 38.27 MB
.........
80: rss: 36.16 MB, heapTotal: 33.80 MB, heapUsed: 29.11 MB
.........
90: rss: 35.14 MB, heapTotal: 32.78 MB, heapUsed: 25.91 MB
.........
100: rss: 33.27 MB, heapTotal: 30.96 MB, heapUsed: 22.83 MB
collections teardown done
time elapsed: 115373ms
closing code: 0
total time elapsed: 240827ms
```

**Comment by blah238** [comment]
I bumped PETS_NUMBER to 1000:

```
$ node test/load


-------------------------------------------------------------------------------------------
                                     sails-postgresql
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
closing code: 3


-------------------------------------------------------------------------------------------
                                     sails-memory
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 74.66 MB, heapTotal: 71.28 MB, heapUsed: 55.16 MB
.........
20: rss: 74.66 MB, heapTotal: 71.28 MB, heapUsed: 55.16 MB
.........
30: rss: 74.66 MB, heapTotal: 71.28 MB, heapUsed: 55.17 MB
.........
40: rss: 74.64 MB, heapTotal: 71.28 MB, heapUsed: 55.17 MB
.........
50: rss: 74.64 MB, heapTotal: 71.28 MB, heapUsed: 55.17 MB
.........
60: rss: 74.64 MB, heapTotal: 71.28 MB, heapUsed: 55.17 MB
.........
70: rss: 74.64 MB, heapTotal: 71.28 MB, heapUsed: 55.17 MB
.........
80: rss: 74.64 MB, heapTotal: 71.28 MB, heapUsed: 55.17 MB
.........
90: rss: 74.64 MB, heapTotal: 71.28 MB, heapUsed: 55.17 MB
.........
100: rss: 74.64 MB, heapTotal: 71.28 MB, heapUsed: 55.17 MB
collections teardown done
time elapsed: 135249ms
closing code: 0


-------------------------------------------------------------------------------------------
                                     sails-disk
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 72.42 MB, heapTotal: 69.14 MB, heapUsed: 57.52 MB
.........
20: rss: 72.42 MB, heapTotal: 69.14 MB, heapUsed: 57.52 MB
.........
30: rss: 72.42 MB, heapTotal: 69.14 MB, heapUsed: 57.52 MB
.........
40: rss: 72.46 MB, heapTotal: 69.14 MB, heapUsed: 57.53 MB
.........
50: rss: 72.46 MB, heapTotal: 69.14 MB, heapUsed: 57.53 MB
.........
60: rss: 72.46 MB, heapTotal: 69.14 MB, heapUsed: 57.53 MB
.........
70: rss: 72.46 MB, heapTotal: 69.14 MB, heapUsed: 57.53 MB
.........
80: rss: 72.46 MB, heapTotal: 69.14 MB, heapUsed: 57.53 MB
.........
90: rss: 72.46 MB, heapTotal: 69.14 MB, heapUsed: 57.53 MB
.........
100: rss: 72.46 MB, heapTotal: 69.14 MB, heapUsed: 57.53 MB
collections teardown done
time elapsed: 137010ms
closing code: 0


-------------------------------------------------------------------------------------------
                                     sails-mysql
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
closing code: 3
total time elapsed: 1083178ms
```

I'm seeing about 700MB usage (with sails-disk and sails-memory as well with 1000 pets. The sails-disk and sails-mysql tests seemed to have crashed out around 1GB.

**Comment by dmarcelino** [comment]
I've just noticed an awful bug in my script, the MB rounding is off by a factor of 10x. 18MB actually means 180MB, just pushed a fix.

**Comment by dmarcelino** [comment]
Looking at the results for 100 pets, in **Travis** node **0.10** all adapters consume more or less the same memory, around **150 MB**, with Mongo reaching 180 MB. I see the same in my mac. Travis build: https://travis-ci.org/balderdashy/waterline-adapter-tests/jobs/56608731#L685

On **windows** (@blah238's machine), the memory and disk adapters take the same memory but SQL adapters are consuming **more than twice** the memory. I don't think waterline has any code differences between OSes so I wonder if it's related to the drivers...

More confusing, on **Travis** node **0.11**,  the memory and disk adapters take around **250 MB** while the SQL adapters reach **740 MB**!

In table format (100 pets):

Machine | Node | memory and disk adapters | SQL Adapters | Reference
------------|---------|-------------------------------------|--------------------|-------
Travis    | 0.10    | ~ 150 MB                              | ~ 150 MB        | [build 19.2](https://travis-ci.org/balderdashy/waterline-adapter-tests/jobs/56608731#L685)
Travis    | 0.11 | ~ 270 MB                              | ~ 740 MB        | [build 19.1](https://travis-ci.org/balderdashy/waterline-adapter-tests/jobs/56608730#L688)
@blah238    | 0.10 | ~ 150 MB                              | ~ 380 MB        | [comment](https://github.com/balderdashy/sails-postgresql/issues/141#issuecomment-88184145)

**Comment by blah238** [comment]
Actually on this machine I was running Node 0.12 for the previous results. I've re-run it with 0.10 (and the off-by-10x fix):

```
$ node test/load


-------------------------------------------------------------------------------------------
                                     sails-postgresql
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 93.78 MB, heapTotal: 79.62 MB, heapUsed: 57.14 MB
.........
20: rss: 89.99 MB, heapTotal: 76.60 MB, heapUsed: 55.98 MB
.........
30: rss: 89.15 MB, heapTotal: 76.58 MB, heapUsed: 60.12 MB
.........
40: rss: 76.83 MB, heapTotal: 63.95 MB, heapUsed: 56.72 MB
.........
50: rss: 76.68 MB, heapTotal: 63.96 MB, heapUsed: 55.48 MB
.........
60: rss: 75.46 MB, heapTotal: 62.94 MB, heapUsed: 56.08 MB
.........
70: rss: 74.47 MB, heapTotal: 59.90 MB, heapUsed: 55.32 MB
.........
80: rss: 75.06 MB, heapTotal: 62.94 MB, heapUsed: 55.98 MB
.........
90: rss: 75.80 MB, heapTotal: 62.93 MB, heapUsed: 58.24 MB
.........
100: rss: 72.86 MB, heapTotal: 60.89 MB, heapUsed: 56.35 MB
collections teardown done
closing code: 0


-------------------------------------------------------------------------------------------
                                     sails-memory
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 78.68 MB, heapTotal: 68.50 MB, heapUsed: 55.15 MB
.........
20: rss: 86.59 MB, heapTotal: 76.60 MB, heapUsed: 62.45 MB
.........
30: rss: 91.01 MB, heapTotal: 79.63 MB, heapUsed: 49.79 MB
.........
40: rss: 91.28 MB, heapTotal: 79.63 MB, heapUsed: 57.12 MB
.........
50: rss: 92.27 MB, heapTotal: 81.66 MB, heapUsed: 70.78 MB
.........
60: rss: 95.92 MB, heapTotal: 84.70 MB, heapUsed: 36.53 MB
.........
70: rss: 96.75 MB, heapTotal: 84.70 MB, heapUsed: 44.21 MB
.........
80: rss: 82.35 MB, heapTotal: 70.50 MB, heapUsed: 29.18 MB
.........
90: rss: 66.01 MB, heapTotal: 55.28 MB, heapUsed: 17.51 MB
.........
100: rss: 55.40 MB, heapTotal: 45.13 MB, heapUsed: 15.34 MB
collections teardown done
closing code: 0


-------------------------------------------------------------------------------------------
                                     sails-disk
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 78.70 MB, heapTotal: 68.50 MB, heapUsed: 52.59 MB
.........
20: rss: 83.79 MB, heapTotal: 73.55 MB, heapUsed: 46.75 MB
.........
30: rss: 84.98 MB, heapTotal: 73.55 MB, heapUsed: 50.95 MB
.........
40: rss: 86.19 MB, heapTotal: 75.57 MB, heapUsed: 58.27 MB
.........
50: rss: 89.97 MB, heapTotal: 78.59 MB, heapUsed: 36.33 MB
.........
60: rss: 90.80 MB, heapTotal: 78.59 MB, heapUsed: 50.28 MB
.........
70: rss: 85.58 MB, heapTotal: 73.55 MB, heapUsed: 29.88 MB
.........
80: rss: 85.58 MB, heapTotal: 73.55 MB, heapUsed: 33.99 MB
.........
90: rss: 69.10 MB, heapTotal: 57.31 MB, heapUsed: 22.62 MB
........closing code: 0


-------------------------------------------------------------------------------------------
                                     sails-mysql
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 96.44 MB, heapTotal: 83.67 MB, heapUsed: 59.77 MB
.........
20: rss: 79.84 MB, heapTotal: 65.98 MB, heapUsed: 56.58 MB
.........
30: rss: 79.16 MB, heapTotal: 65.98 MB, heapUsed: 61.67 MB
.........
40: rss: 80.26 MB, heapTotal: 66.98 MB, heapUsed: 58.87 MB
.........
50: rss: 79.84 MB, heapTotal: 66.98 MB, heapUsed: 62.45 MB
.........
60: rss: 79.95 MB, heapTotal: 66.98 MB, heapUsed: 62.26 MB
.........
70: rss: 79.77 MB, heapTotal: 66.98 MB, heapUsed: 56.17 MB
.........
80: rss: 78.61 MB, heapTotal: 64.96 MB, heapUsed: 56.52 MB
.........
90: rss: 78.46 MB, heapTotal: 64.98 MB, heapUsed: 58.87 MB
.........
100: rss: 78.51 MB, heapTotal: 64.96 MB, heapUsed: 54.12 MB
collections teardown done
closing code: 0
total time elapsed: 266354ms
```

Things still go crazy when PETS_NUMBER=1000, however.

**Comment by dmarcelino** [comment]
With PETS_NUMBER=1000 in my machine sails-memory and sails-disk consume close to 1 GB. Memory actually gets lower and lower within runs.
``` shell
 sails-memory 
---------------------
0: rss: 60.53 MB, heapTotal: 47.93 MB, heapUsed: 23.57 MB
10: rss: 994.05 MB, heapTotal: 971.15 MB, heapUsed: 944.50 MB
20: rss: 926.16 MB, heapTotal: 902.05 MB, heapUsed: 871.68 MB
30: rss: 749.57 MB, heapTotal: 726.30 MB, heapUsed: 681.58 MB
40: rss: 713.81 MB, heapTotal: 694.30 MB, heapUsed: 533.52 MB
50: rss: 652.25 MB, heapTotal: 631.84 MB, heapUsed: 419.59 MB
60: rss: 567.96 MB, heapTotal: 548.24 MB, heapUsed: 346.31 MB
70: rss: 459.19 MB, heapTotal: 440.40 MB, heapUsed: 329.60 MB
80: rss: 445.89 MB, heapTotal: 428.50 MB, heapUsed: 339.28 MB
90: rss: 274.68 MB, heapTotal: 259.33 MB, heapUsed: 124.22 MB
100: rss: 130.04 MB, heapTotal: 117.02 MB, heapUsed: 56.61 MB
```
It seems to me that find.populate just consumes whatever memory it cans.

BTW, if the hits are processed in series (instead of parallel) then the memory taken is around 100 MB.

**Comment by blah238** [comment]
Here is my PETS_NUMBER=1000. I guess the SQL adapters just crash out.

```
$ node test/load


-------------------------------------------------------------------------------------------
                                     sails-postgresql
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
closing code: 3


-------------------------------------------------------------------------------------------
                                     sails-memory
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 522.11 MB, heapTotal: 497.93 MB, heapUsed: 481.25 MB
.........
20: rss: 605.32 MB, heapTotal: 577.89 MB, heapUsed: 561.61 MB
.........
30: rss: 479.05 MB, heapTotal: 453.81 MB, heapUsed: 401.40 MB
.........
40: rss: 431.90 MB, heapTotal: 408.12 MB, heapUsed: 324.60 MB
.........
50: rss: 390.09 MB, heapTotal: 367.48 MB, heapUsed: 272.15 MB
.........
60: rss: 359.67 MB, heapTotal: 338.01 MB, heapUsed: 233.93 MB
.........
70: rss: 326.50 MB, heapTotal: 306.51 MB, heapUsed: 218.74 MB
.........
80: rss: 238.13 MB, heapTotal: 220.25 MB, heapUsed: 145.26 MB
.........
90: rss: 141.86 MB, heapTotal: 126.88 MB, heapUsed: 80.74 MB
.........
100: rss: 47.60 MB, heapTotal: 36.55 MB, heapUsed: 16.49 MB
collections teardown done
closing code: 0


-------------------------------------------------------------------------------------------
                                     sails-disk
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
.........
10: rss: 497.72 MB, heapTotal: 474.51 MB, heapUsed: 458.67 MB
.........
20: rss: 427.73 MB, heapTotal: 405.65 MB, heapUsed: 354.00 MB
.........
30: rss: 474.77 MB, heapTotal: 451.13 MB, heapUsed: 434.32 MB
.........
40: rss: 369.21 MB, heapTotal: 347.71 MB, heapUsed: 304.58 MB
.........
50: rss: 335.31 MB, heapTotal: 316.19 MB, heapUsed: 216.30 MB
.........
60: rss: 347.52 MB, heapTotal: 328.25 MB, heapUsed: 298.60 MB
.........
70: rss: 285.17 MB, heapTotal: 266.36 MB, heapUsed: 242.81 MB
.........
80: rss: 257.12 MB, heapTotal: 238.96 MB, heapUsed: 202.42 MB
.........
90: rss: 167.20 MB, heapTotal: 151.70 MB, heapUsed: 102.29 MB
.........closing code: 0


-------------------------------------------------------------------------------------------
                                     sails-mysql
-------------------------------------------------------------------------------------------

collections initialized
Hitting the db with 100 requests...
closing code: 3
total time elapsed: 1697394ms
```

**Comment by dmarcelino** [comment]
I just did some calculations based on my sails-memory results (node 0.10):

base memory (hits = 0) | max memory | users | pets/user | hits  | memory / populate | memory / populate / pet
-------------------|-------------------|---------|--------------|-------|-------------------------|-----------------------------
36 MB            | 170 MB          | 3       | 100          | 100  |   446 KB                | 4.46 KB
60 MB            | 994 MB          | 3       | 100          | 100  |  3.113 MB                | 3.11 KB

memory/populate = ( max memory - base memory ) / ( users * hits)

Gives me the impression that .populate() is a memory expensive operation.

**Comment by devinivy** [comment]
In the sense that it's leaking memory, yes?  It never settles down?

**Comment by dmarcelino** [comment]
I don't think it's leaking as it comes down, it just seems that each populate consumes lots of memory. Is waterline cloning things during populates?

**Comment by devinivy** [comment]
Yeah, it probably does.  I really expected this to be due to waterline-cursor, which does memory-intensive operations like in-memory joins.  But mysql wouldn't have quite the same issue.  The fact that it's also apparently taxing the CPU is concerning.  Someone is going to have to go deeper to see where memory is being used.  Have either of you used used Chrome's heap inspection tools?  They're very good.  I think it's the next place to go.

@dmarcelino you said that when these requests go in series that there's no apparent leak, yes?  I'm just really trying to figure out whether we need performance improvements or need to find a genuine leak.

**Comment by blah238** [comment]
@devinivy @dmarcelino @tjwebb @particlebanana what do you think about moving this issue over to the waterline repo since it seems to affect multiple (all?) adapters (albeit in different ways)?

@devinivy I was playing around with node-inspector's profile tab yesterday but wasn't really sure what to look for. I probably can't share a heap snapshot for security reasons since there's no telling what's in there.

**Comment by dmarcelino** [comment]
@devinivy, I don't think there is ever a memory leak in any scenario as the memory eventually comes down. It's just that `populate()` loads a lot of things into memory which leads to a big memory footprint. When hitting the DB in series, the max memory is lower which makes sense, as the `populate()` operations never overlap each other.

We will need to profile waterline and figure out why `populate()` is so memory hungry...

**Comment by devinivy** [comment]
First round of looking through this leads me to believe it has something to do with the buffers here:
https://github.com/balderdashy/waterline-cursor/blob/master/cursor/populateBuffers.js#L34 & https://github.com/balderdashy/waterline-cursor/blob/master/index.js#L36

**Comment by blah238** [comment]
As I mentioned over in the comments, PR https://github.com/balderdashy/waterline/pull/829 improves this situation massively (~45x faster, 1/3 the memory consumption). It still should be much faster and less memory-intensive, but in my scenario it's gone from completely unworkable to bearable.

I'll do another round of load tests later on with the patch.

**Comment by devinivy** [comment]
Excellent!  I'm pushing for a release by the end of the week with that PR and other critical changes.  I think we should continue to be skeptical of strategies used in waterline-cursor, and I invite you all to check it out.  Bear in mind, each adapter may override default methods used in waterline-cursor.

**Comment by billyshena** [comment]
Hello guys ! I'm encountering the same performance issues using .populate (just a single field), eating up the memory & CPU usage. I'm using the sails-postgresql 0.10.15 adapter.
Is there any fix released for those perf issues?

Best,

**Comment by sailsbot** [comment]
Thanks for posting, @blah238.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by mikermcneil** [comment]
For future reference, there's more information and tips on diagnosing memory leaks in Node apps, and on reproducing+reporting suspected memory leaks in Sails here: https://github.com/balderdashy/sails/issues/3782#issuecomment-238679052
